sklearn/decomposition/_online_lda_fast.pyx

```diff
@@ -39,23 +54,40 @@ def _dirichlet_expectation_1d(np.ndarray[ndim=1, dtype=np.float64_t] doc_topic,
         out[:] = np.exp(psi(doc_topic) - psi(np.sum(doc_topic)))
     """
 
-    cdef np.float64_t dt, psi_total, total
+    cdef numpy_float_t dt, psi_total, total
+    cdef np.float64_t dt_64, psi_total_64, total_64
     cdef np.npy_intp i, size
 
     size = doc_topic.shape[0]
 
-    total = 0.0
-    for i in range(size):
-        dt = doc_topic[i] + doc_topic_prior
-        doc_topic[i] = dt
-        total += dt
-    psi_total = psi(total)
-
-    for i in range(size):
-        out[i] = exp(psi(doc_topic[i]) - psi_total)
-
 
-def _dirichlet_expectation_2d(np.ndarray[ndim=2, dtype=np.float64_t] arr):
+    if numpy_float_t == numpy_float2_t:
```

Comment:
We can use fused types directly here instead of conditionally on the dtype:

```python
    cdef numpy_float_t dt
    cdef numpy_float2_t total, psi_total
    cdef np.npy_intp i, size

    size = doc_topic.shape[0]

    total = 0.0
    for i in range(size):
        dt = doc_topic[i] + doc_topic_prior
        doc_topic[i] = dt
        total += dt
    psi_total = psi(total)

    for i in range(size):
        out[i] = exp(psi(doc_topic[i]) - psi_total)
```

Although, I think all the total should be float64, since it has more maximum capacity. In other words:

```python
import numpy as np

np.finfo(np.float32).max
# 3.402e+38

np.finfo(np.float64).max
# 1.79769e+308
```

When summing up many `float32s`, the sum can overflow if the total is over `3.4e+38`. To prevent this we can always use `float64` for the total. In other words something like this:

```python
    cdef numpy_float_t dt
    cdef np.float64_t total, psi_total
    cdef np.npy_intp i, size

    size = doc_topic.shape[0]

    total = 0.0
    for i in range(size):
        dt = doc_topic[i] + doc_topic_prior
        doc_topic[i] = dt
        total += dt
    psi_total = psi(total)

    for i in range(size):
        out[i] = exp(psi(doc_topic[i]) - psi_total)
```

Reply:
Thank you for the suggestion.
Yes, this looks better. 
I will Change to use float64 for accumulation.